src/cursor/core_cursor.ts

```diff
@@ -3,139 +3,189 @@ import { ReadPreference } from '../read_preference';
+  transform?(doc: Document): Document;
+}
+
+export interface InternalCursorState extends BSONSerializeOptions {
+  postBatchResumeToken?: ResumeToken;
+  batchSize: number;
+  cmd: Document;
+  currentLimit: number;
+  cursorId: Long | null;
+  lastCursorId?: Long | null;
+  cursorIndex: number;
+  dead: boolean;
+  killed: boolean;
+  init: boolean;
+  notified: boolean;
+  documents: Document[];
+  limit: number;
+  operationTime?: OperationTime;
+  reconnect?: boolean;
+  session?: ClientSession;
+  skip: number;
+  streamOptions?: StreamOptions;
+  transforms?: TransformFunctions;
+  raw?: boolean;
+}
 
 // Possible states for a cursor
-const CursorState = {
-  INIT: 0,
-  OPEN: 1,
-  CLOSED: 2,
-  GET_MORE: 3
-};
-
-interface CoreCursor {
-  close(options?: any, callback?: Callback): Promise<void> | void;
+export enum CursorState {
+  INIT = 0,
+  OPEN = 1,
+  CLOSED = 2,
+  GET_MORE = 3
 }
 
-/**
- * This is a cursor results callback
- *
- * @callback resultCallback
- * @param {error} error An error object. Set to null if no error present
- * @param {object} document
- */
+export interface CoreCursorOptions extends BSONSerializeOptions {
+  noCursorTimeout?: boolean;
+  tailable?: boolean;
+  raw?: boolean;
+  hint?: Hint;
+  disconnectHandler?: DisconnectHandler;
+  limit?: number;
+  skip?: number;
+  session?: ClientSession;
+  readPreference?: ReadPreference;
+  /** The number of documents to return per batch. See {@link https://docs.mongodb.com/manual/reference/command/find/| find command documentation} and {@link https://docs.mongodb.com/manual/reference/command/aggregate|aggregation documentation}. */
+  batchSize?: number;
+  /** Initial documents list for cursor */
+  documents?: Document[];
+  /** Transform function */
+  transforms?: TransformFunctions;
+}
 
 /**
  * The **CoreCursor** class is an internal class that embodies a cursor on MongoDB
  * allowing for iteration over the results returned from the underlying query.
  *
  * **CURSORS Cannot directly be instantiated**
- *
- * @property {number} cursorBatchSize The current cursorBatchSize for the cursor
- * @property {number} cursorLimit The current cursorLimit for the cursor
- * @property {number} cursorSkip The current cursorSkip for the cursor
  */
-class CoreCursor extends Readable {
-  operation: any;
-  pool: any;
-  server: any;
-  disconnectHandler: any;
-  ns: any;
-  namespace: any;
-  cmd: any;
-  options: any;
-  topology: any;
+export class CoreCursor<
+  T extends CoreCursorOptions = CoreCursorOptions,
+  O extends OperationBase = OperationBase
+> extends Readable {
+  operation?: O;
+  pool!: ConnectionPool;
+  server!: Server;
+  disconnectHandler?: DisconnectHandler;
+  ns: string;
+  namespace: MongoDBNamespace;
+  cmd: Document;
+  options: T;
+  topology: Topology;
   cursorState: InternalCursorState;
-  logger: any;
-  query: any;
-  s: any;
+  logger: Logger;
+  query?: Document;
+  s!: CoreCursorPrivate;
```

Comment:
Bit of a mess but only subclasses assign and make their s property but CoreCursor doesn't. I think I could move the logic of making the s prop into core cursor. The hardest part is avoiding optional `?:` on things that will always be defined because the whole s object needs to be declared at once.

Reply:
ok maybe we can leave this alone for now then